algebras/algebra/src/main/scala/endpoints4s/algebra/WebSockets.scala

```diff
@@ -0,0 +1,64 @@
+package endpoints4s.algebra
+
+import endpoints4s.Tupler
+
+trait WebSockets extends WebsocketClientStream with WebsocketServerStream with Urls {
+
+  type WebSocket[A, B]
+
+  def webSocket[URLC, CSC, In, Out](
+      url: Url[URLC],
+      clientStream: WebsocketClientStream[CSC] = emptyWebsocketClientStream,
+      serverStream: WebsocketServerStream[Out] = emptyWebsocketServerStream
+  )(implicit
+      tupler: Tupler.Aux[URLC, CSC, In]
+  ): WebSocket[In, Out]
+}
+
+trait WebSocketsWithUpgrade
+    extends WebSockets
+    with WebSocketUpgradeRequest
+    with WebSocketUpgradeResponse {
+
+  type WebSocketWithUpgrade[A, B]
+
+  def webSocketWithUpgrade[URLC, CSC, URHC, SSS, URHS, URLCAndCSCTupled, In, Out](
```

Comment:
I am not familiar enough with the upgrade thing, but since you are just using a standard `RequestHeaders` parameter, I wonder if we should just replace the `webSocket` constructor with this one? Why do we want to have both?

Reply:
Reading your last comment again, now I think I understand: in the case of web browsers, there would be no possibility of manipulating the request or response headers?

Reply:
Yes, the exact reasoning was provided in that comment - there is no _explicit_ way to interact with the headers, but browsers can _implicitly_ pass some headers. Based on reasons provided, does this separation make sense to you?

Reply:
Yes.

Reply:
Then, indeed, if we want to provide `webSocketWithUpgrade` (for those who can implement it), it has to be in a separate trait. Thank you for the explanation!